src/app/ClusterData.h

```diff
@@ -0,0 +1,382 @@
+     */
+    uint64_t GetTag(AttributePathHandle aHandle) const;
+
+    /**
+     * Returns the parent handle of a given child path handle. Dictionary keys in the handle are preserved in the case where the
+     * parent handle is also a dictionary element.
+     *
+     * @retval AttributePathHandle   Handle of the parent.
+     */
+    AttributePathHandle GetParent(AttributePathHandle aHandle) const;
+
+    /**
+     * Returns the first child handle associated with a particular parent.
+     *
+     * @retval AttributePathHandle   Handle of the first child.
+     */
+    AttributePathHandle GetFirstChild(AttributePathHandle aParentHandle) const;
+
+    /**
+     * Given a handle to an existing child, returns the next child handle associated with a particular parent.
+     *
+     * @retval AttributePathHandle   Handle of the next child.
+     */
+    AttributePathHandle GetNextChild(AttributePathHandle aParentId, AttributePathHandle aChildHandle) const;
+
+    /**
+     * Returns the cluster id of the associated Cluster.
+     *
+     * @retval Cluster id
+     */
+    ClusterId GetClusterId(void) const;
+
+    /**
+     * Checks if a given handle is a child of another handle. This can be an in-direct parent.
+     *
+     * @retval bool
+     */
+    bool IsParent(AttributePathHandle aChildHandle, AttributePathHandle aParentHandle) const;
+
+    AttributePathHandle GetChildHandle(AttributePathHandle aParentHandle, uint8_t aContextTag) const;
+
+    const Schema mSchema;
+};
+
+/*
+ * @class  ClusterDataSink
+ *
+ * @brief  Base abstract class that represents a particular instance of a Cluster on a specific external resource (client).
+ * Application developers are expected to subclass this to make a concrete sink that ingests data received from servers.
+ *
+ *         It takes in a pointer to a schema that it then uses to help decipher incoming TLV data from a publisher and invoke the
+ *         relevant data setter calls to pass the data up to subclasses.
+ */
+class ClusterDataSink : protected ClusterSchemaEngine::ISetDataDelegate
+{
+public:
+    ClusterDataSink(const ClusterSchemaEngine * apEngine);
+    virtual ~ClusterDataSink() {}
+    const ClusterSchemaEngine * GetSchemaEngine(void) const { return mpSchemaEngine; }
+
+    /**
+     * Given a reader that points to a data element conformant to a schema bound to this object, this method processes that data and
+     * invokes the relevant SetLeafData call below for all leaf items in the buffer.
+     *
+     *
+     * @retval #CHIP_NO_ERROR On success.
+     * @retval other           Encountered errors writing out the data.
+     */
+    CHIP_ERROR StoreDataElement(AttributePathHandle aHandle, TLV::TLVReader & aReader);
+
+    /**
+     * Retrieves the current version of the data that resides in this sink.
+     */
+    DataVersion GetVersion(void) const { return mVersion; }
+
+    /**
+     * Returns a boolean value that determines whether the version is valid.
```

Comment:
delete this entire comment: it just restates `IsVersionValid` in a longer sentence. It would be nice if it would describe what makes a version valid and it is a version of what (when is mHasValidVersion set?).